src/main/java/com/uber/cadence/internal/common/WorkflowExecutionUtils.java

```diff
@@ -191,31 +194,46 @@ public static HistoryEvent getInstanceCloseEvent(
     long start = System.currentTimeMillis();
     HistoryEvent event;
     do {
+      if (timeout != 0 && System.currentTimeMillis() - start > unit.toMillis(timeout)) {
+        throw new TimeoutException(
+            "WorkflowId="
+                + workflowExecution.getWorkflowId()
+                + ", runId="
+                + workflowExecution.getRunId()
+                + ", timeout="
+                + timeout
+                + ", unit="
+                + unit);
+      }
+
       GetWorkflowExecutionHistoryRequest r = new GetWorkflowExecutionHistoryRequest();
       r.setDomain(domain);
       r.setExecution(workflowExecution);
       r.setHistoryEventFilterType(HistoryEventFilterType.CLOSE_EVENT);
       r.setNextPageToken(pageToken);
       r.setWaitForNewEvent(true);
+      r.setSkipArchival(true);
       try {
         response =
             Retryer.retryWithResult(retryParameters, () -> service.GetWorkflowExecutionHistory(r));
       } catch (EntityNotExistsError e) {
+        if (e.activeCluster != null
+            && e.currentCluster != null
+            && !e.activeCluster.equals(e.currentCluster)) {
+          // Current cluster is passive cluster. Execution might not exist because of replication
+          // lag. Wait for a little bit and retry.
+          try {
+            Thread.sleep(ENTITY_NOT_EXIST_RETRY_WAIT_MILLIS);
+          } catch (InterruptedException ie) {
+            // Throw entity not exist here.
+            throw e;
+          }
+          continue;
```

Comment:
With your change, we no longer throw an `EntityNotExistsError` when the passive cluster is queried; except the `InterruptedException` case above. In all other cases where we simply `continue` and retry, we would eventually throw a `TimeoutException`. I find that a bit misleading.

I'd consider throwing the most recent `EntityNotExistsError` instead of a new `TimeoutException` when we give up on retries due to timeout. Is there any reason for not doing that?

Reply:
Done.